producer.go

```diff
@@ -95,12 +152,24 @@ func (p *segmentProducer) configureSecurity(cfg Config) error {
 	return nil
 }
 
-func (p *segmentProducer) Send(key []byte, payload []byte) error {
+func (p *segmentProducer) Send(key []byte, payload []byte, id []byte, ackFunc sdk.AckFunc) error {
+	// accessed also in onMessageDelivery, which is
+	// calling and removing the ack functions
+	p.m.Lock()
+	p.ackFuncs[string(id)] = ackFunc
+	p.m.Unlock()
+
 	err := p.writer.WriteMessages(
```

Comment:
If the write failed we need to remove `ackFunc` from the map and negatively acknowledge the record by calling `ackFunc` with the error. The problem is of course that we can't decide to do that here since the destination will retry the write 10 times. Either we need to push retries into the producer or push ack handling up to the destination.

Reply:
Good catch!